File: src/main/java/org/treblereel/gwt/three4g/core/BufferAttribute.java
package org.treblereel.gwt.three4g.core;

import elemental2.core.JsArray;
import elemental2.core.TypedArray;
import jsinterop.annotations.JsConstructor;
import jsinterop.annotations.JsType;
import org.treblereel.gwt.three4g.math.Color;
import org.treblereel.gwt.three4g.math.Vector2;
import org.treblereel.gwt.three4g.math.Vector3;
import org.treblereel.gwt.three4g.math.Vector4;

/**
 * This class stores data for an attribute (such as vertex positions, face indices, normals, colors, UVs, and any custom
 * attributes ) associated with a BufferGeometry, which allows for more efficient passing of data to the GPU. See that page
 * for details and a usage example.
 * <p>
 * Data is stored as vectors of any length (defined by itemSize), and in general in the methods outlined below if passing
 * in an index, this is automatically multiplied by the vector length.
 *
 * @author Dmitrii Tikhomirov <chani@me.com>
 * Created by treblereel on 3/1/18.
 */
@JsType(isNative = true, namespace = "THREE")
public class BufferAttribute {

    /**
     * The array holding data stored in the buffer.
     */
    public TypedArray array;

    /**
     * Stores the array's length divided by the itemSize.
     * If the buffer is storing a 3-component vector (such as a position, normal, or color), then this will count the number of such vectors stored.
     */
    public int count;

    /**
     * Whether the buffer is dynamic or not. Default is false.
     * If false, the GPU is informed that contents of the buffer are likely to be used often and not change often. This corresponds to the gl.STATIC_DRAW flag.
     * if true, the GPU is informed that contents of the buffer are likely to be used often and change often. This corresponds to the gl.DYNAMIC_DRAW flag.
     */
    public boolean dynamic;

    /**
     * Used to check whether this or derived classes are BufferAttributes. Default is true.
     * <p>
     * You should not change this, as it used internally for optimisation.
     */
    public boolean isBufferAttribute;

    /**
     * The length of vectors that are being stored in the array.
     */
    public int itemSize;

    /**
     * Optional name for this attribute instance. Default is an empty string.
     */
    public String name;

    /**
     * Flag to indicate that this attribute has changed and should be re-sent to the GPU. Set this to true when you modify the value of the array.
     * <p>
     * Setting this to true also increments the version.
     */
    public boolean needsUpdate;


    /**
     * Indicates how the underlying data in the buffer maps to the values in the GLSL shader code. See the constructor above for details.
     */
    public boolean normalized;

    /**
     * A callback function that is executed after the Renderer has transfered the attribute array data to the GPU.
     */
    public OnUploadCallback onUploadCallback;

    /**
     * Object containing:
     * offset: Default is 0. Position at whcih to start update.
     * count: Default is -1, which means don't use update ranges.
     * <p>
     * This can be used to only update some components of stored vectors (for example, just the component related to color).
     */
    public UpdateRange updateRange;

    /**
     * A version number, incremented every time the needsUpdate property is set to true.
     */
    public int version;

    BufferAttribute() {

    }

    /**
     * @param array    -- Must be a TypedArray. Used to instantiate the buffer.
     *                 This array should have
     *                 itemSize * numVertices elements, where numVertices is the number of vertices in the associated BufferGeometry.
     * @param itemSize -- the number of values of the array that should be associated with a particular vertex. For instance, if this
     *                 attribute is storing a 3-component vector (such as a position, normal, or color), then itemSize should be 3.
     */
    @JsConstructor
    public BufferAttribute(TypedArray array, int itemSize) {

    }

    /**
     * @param array      -- Must be a TypedArray. Used to instantiate the buffer.
     *                   This array should have
     *                   itemSize * numVertices elements, where numVertices is the number of vertices in the associated BufferGeometry.
     * @param itemSize   -- the number of values of the array that should be associated with a particular vertex. For instance, if this
     *                   attribute is storing a 3-component vector (such as a position, normal, or color), then itemSize should be 3.
     * @param normalized -- (optional) Indicates how the underlying data in the buffer maps to the values in the GLSL code.
     *                   For instance, if array is an instance of UInt16Array, and normalized is true, the values 0 - +65535
     *                   in the array data will be mapped to 0.0f - +1.0f in the GLSL attribute. An Int16Array (signed)
     *                   would map from -32767 - +32767 to -1.0f - +1.0f. If normalized is false, the values will be converted
     *                   to floats which contain the exact value, i.e. 32767 becomes 32767.0f.
     */
    @JsConstructor
    public BufferAttribute(TypedArray array, int itemSize, boolean normalized) {

    }

    /**
     * Return a copy of this bufferAttribute.
     *
     * @return
     */
    public native BufferAttribute clone();

    /**
     * Copy the array given here (which can be a normal array or TypedArray) into array.
     * <p>
     * See TypedArray.set for notes on requirements if copying a TypedArray.
     *
     * @param array
     * @return
     */
    public native BufferAttribute copyArray(TypedArray array);

    /**
     * Copy the array given here (which can be a normal array or TypedArray) into array.
     *
     * @param array
     * @return
     */
    public native BufferAttribute copyArray(JsArray array);

    /**
     * Copy a vector from bufferAttribute[index2] to array[index1].
     *
     * @param index1
     * @param bufferAttribute
     * @param index2
     */
    public native void copyAt(int index1, BufferAttribute bufferAttribute, int index2);

    /**
     * Copy an array representing RGB color values into array.
     *
     * @param colors
     */
    public native BufferAttribute copyColorsArray(Color[] colors);

    /**
     * Copy an array representing Vector2s into array.
     *
     * @param vectors
     * @return
     */
    public native BufferAttribute copyVector2sArray(Vector2[] vectors);

    /**
     * Copy an array representing Vector2s into array.
     *
     * @param vectors
     * @return
     */
    public native BufferAttribute copyVector3sArray(Vector3[] vectors);

    /**
     * Copy an array representing Vector4s into array.
     *
     * @param vectors
     * @return
     */
    public native BufferAttribute copyVector4sArray(Vector4[] vectors);

    /**
     * Returns the x component of the vector at the given index.
     *
     * @param index
     * @return
     */
    public native Object getX(int index);

    /**
     * Returns the y component of the vector at the given index.
     *
     * @param index
     * @return
     */
    public native Object getY(int index);

    /**
     * Returns the z component of the vector at the given index.
     *
     * @param index
     * @return
     */
    public native Object getZ(int index);

    /**
     * Sets the value of the onUploadCallback property.
     * <p>
     * In the WebGL / Buffergeometry this is used to free memory after the buffer has been	transfered to the GPU.
     *
     * @param callback
     */
    public native void onUpload(OnUploadCallback callback);

    /**
     * @param value -- TypedArray from which to copy values.
     *              <p>
     *              In particular, see that page for requirements on value being a TypedArray.
     * @return
     */
    public native BufferAttribute set(TypedArray value);

    /**
     * @param value  -- TypedArray from which to copy values.
     * @param offset -- index of the array at which to start copying.
     *               Calls	TypedArray.set( value, offset ) on the array.
     *               <p>
     *               In particular, see that page for requirements on value being a TypedArray.
     * @return
     */
    public native BufferAttribute set(TypedArray value, int offset);

    /**
     * @param value -- an Array or JsArray from which to copy values.
     *              <p>
     *              In particular, see that page for requirements on value being a TypedArray.
     * @return
     */
    public native BufferAttribute set(JsArray value);

    /**
     * @param value  -- JsArray from which to copy values.
     * @param offset -- index of the array at which to start copying.
     *               Calls TypedArray.set( value, offset ) on the array.
     *               <p>
     *               In particular, see that page for requirements on value being a TypedArray.
     * @return
     */
    public native BufferAttribute set(JsArray value, int offset);


    /**
     * array to the TypedArray passed in here.
     * <p>
     * After setting the array, needsUpdate should be set to true.
     *
     * @param array
     * @return
     */
    public native BufferAttribute setArray(TypedArray array);

    /**
     * Set dynamic to value.
     *
     * @param value
     * @return
     */
    public native BufferAttribute setDynamic(boolean value);

    /**
     * Sets the x component of the vector at the given index.
     *
     * @param x
     * @return
     */
    public native BufferAttribute setX(int index, Object x);

    /**
     * Sets the y component of the vector at the given index.
     *
     * @param y
     * @return
     */
    public native BufferAttribute setY(int index, Object y);

    /**
     * Sets the z component of the vector at the given index.
     *
     * @param z
     * @return
     */
    public native BufferAttribute setZ(int index, Object z);

    /**
     * Sets the w component of the vector at the given index.
     *
     * @param w
     * @return
     */
    public native BufferAttribute setW(int index, Object w);

    /**
     * Sets the x and y components of the vector at the given index.
     *
     * @param x
     * @param y
     * @return
     */
    public native BufferAttribute setXY(int index, Object x, Object y);

    /**
     * Sets the x, y and z components of the vector at the given index.
     *
     * @param x
     * @param y
     * @param z
     * @return
     */
    public native BufferAttribute setXYZ(int index, Object x, Object y, Object z);

    /**
     * Sets the x, y, z and w components of the vector at the given index.
     *
     * @param x
     * @param y
     * @param z
     * @param w
     * @return
     */
    public native BufferAttribute setXYZW(int index, Object x, Object y, Object z, Object w);
}

File: core/src/main/resources/org/treblereel/gwt/three4g/core/BufferAttribute.js
import { Vector4 } from '../math/Vector4.js';
import { Vector3 } from '../math/Vector3.js';
import { Vector2 } from '../math/Vector2.js';
import { Color } from '../math/Color.js';
import { StaticDrawUsage } from '../constants.js';
import { Matrix3 } from '../math/Matrix3.js';

import { Matrix4 } from '../math/Matrix4.js';


const _vector = new Vector3();
const _vector2 = new Vector2();

class BufferAttribute {


	/**
	 *
	 * @param {TypedArray|Array<number>} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		if (Array.isArray(array)) {

			throw new TypeError('THREE.BufferAttribute: array should be a Typed Array.');

		}
		/**@type {string} */
		this.name = '';
		/** @type {TypedArray|Array<number>} */
		this.array = array;
		/** @type {number} */
		this.itemSize = itemSize;
		/** @type {number} */
		this.count = array !== undefined ? array.length / itemSize : 0;
		/** @type {boolean} */
		this.normalized = normalized === true;

		/** @type {number} */ 
		this.usage = StaticDrawUsage;
		this.updateRange = { offset: 0, count: - 1 };

		/** @type {number} */
		this.version = 0;
		/** @type {boolean} */
		this.isBufferAttribute = true;
		/** @type {boolean} */
		this.isGLBufferAttribute;

	}

	/**
	 * @param {boolean} value
	 */
	set needsUpdate(value) {

		if (value === true) this.version++;

	}


	onUploadCallback() { }

	/**
	 * @param {number} value
	 * @return {BufferAttribute} 
	 */
	setUsage(value) {

		this.usage = value;

		return this;

	}

	/**
	 * @param {BufferAttribute} source
	 * @return {BufferAttribute} 
	 */
	copy(source) {

		this.name = source.name;
		this.array = new source.array.constructor(source.array);
		this.itemSize = source.itemSize;
		this.count = source.count;
		this.normalized = source.normalized;

		this.usage = source.usage;

		return this;

	}

	/**
	 * @param {number} index1 
	 * @param {BufferAttribute} attribute 
	 * @param {number} index2 
	 * @return {BufferAttribute}
	 */
	copyAt(index1, attribute, index2) {

		index1 *= this.itemSize;
		index2 *= attribute.itemSize;

		for (let i = 0, l = this.itemSize; i < l; i++) {

			this.array[index1 + i] = attribute.array[index2 + i];

		}

		return this;

	}

	/**
	 * @param {!Array<number>} array 
 	 * @return {BufferAttribute}
	 */
	copyArray(array) {

		this.array.set(array);

		return this;

	}

	/**
	 * @param {Array<{ r: number, g: number, b: number }>} colors 
 	 * @return {BufferAttribute}
	 */
	copyColorsArray(colors) {

		const array = this.array;
		let offset = 0;

		for (let i = 0, l = colors.length; i < l; i++) {

			let color = colors[i];

			if (color === undefined) {

				console.warn('THREE.BufferAttribute.copyColorsArray(): color is undefined', i);
				color = new Color();

			}

			array[offset++] = color.r;
			array[offset++] = color.g;
			array[offset++] = color.b;

		}

		return this;

	}

	/**
	 * @param {Array<Vector2>} vectors 
 	 * @return {BufferAttribute}
	 */
	copyVector2sArray(vectors) {

		const array = this.array;
		let offset = 0;

		for (let i = 0, l = vectors.length; i < l; i++) {

			let vector = vectors[i];

			if (vector === undefined) {

				console.warn('THREE.BufferAttribute.copyVector2sArray(): vector is undefined', i);
				vector = new Vector2();

			}

			array[offset++] = vector.x;
			array[offset++] = vector.y;

		}

		return this;

	}

	/**
	 * @param {Array<Vector3>} vectors 
 	 * @return {BufferAttribute}
	 */
	copyVector3sArray(vectors) {

		const array = this.array;
		let offset = 0;

		for (let i = 0, l = vectors.length; i < l; i++) {

			let vector = vectors[i];

			if (vector === undefined) {

				console.warn('THREE.BufferAttribute.copyVector3sArray(): vector is undefined', i);
				vector = new Vector3();

			}

			array[offset++] = vector.x;
			array[offset++] = vector.y;
			array[offset++] = vector.z;

		}

		return this;

	}

	/**
	 * @param {Array<Vector4>} vectors 
 	 * @return {BufferAttribute}
	 */
	copyVector4sArray(vectors) {

		const array = this.array;
		let offset = 0;

		for (let i = 0, l = vectors.length; i < l; i++) {

			/** @type {Vector4} */
			let vector = vectors[i];

			if (vector === undefined) {

				console.warn('THREE.BufferAttribute.copyVector4sArray(): vector is undefined', i);
				vector = new Vector4();

			}

			array[offset++] = vector.x;
			array[offset++] = vector.y;
			array[offset++] = vector.z;
			array[offset++] = vector.w;

		}

		return this;

	}

	/**
	 * @param {Matrix3} m 
 	 * @return {BufferAttribute}
	 * @suppress {checkTypes} 
	 */
	applyMatrix3(m) {

		if (this.itemSize === 2) {

			for (let i = 0, l = this.count; i < l; i++) {

				_vector2.fromBufferAttribute(this, i);
				_vector2.applyMatrix3(m);

				this.setXY(i, _vector2.x, _vector2.y);

			}

		} else if (this.itemSize === 3) {

			for (let i = 0, l = this.count; i < l; i++) {

				_vector.fromBufferAttribute(this, i);
				_vector.applyMatrix3(m);

				this.setXYZ(i, _vector.x, _vector.y, _vector.z);

			}

		}

		return this;

	}

	/**
	 * @param {Matrix4} m 
 	 * @return {BufferAttribute}
	 * @suppress {checkTypes} 
	 */
	applyMatrix4(m) {

		for (let i = 0, l = this.count; i < l; i++) {

			_vector.x = this.getX(i);
			_vector.y = this.getY(i);
			_vector.z = this.getZ(i);

			_vector.applyMatrix4(m);

			this.setXYZ(i, _vector.x, _vector.y, _vector.z);

		}

		return this;

	}

	/**
	 * @param {Matrix3} m 
 	 * @return {BufferAttribute}
	 * @suppress {checkTypes} 
	 */
	applyNormalMatrix(m) {

		for (let i = 0, l = this.count; i < l; i++) {

			_vector.x = this.getX(i);
			_vector.y = this.getY(i);
			_vector.z = this.getZ(i);

			_vector.applyNormalMatrix(m);

			this.setXYZ(i, _vector.x, _vector.y, _vector.z);

		}

		return this;

	}

	/**
	 * @param {Matrix4} m 
 	 * @return {BufferAttribute}
	 * @suppress {checkTypes}  
	 */
	transformDirection(m) {

		for (let i = 0, l = this.count; i < l; i++) {

			_vector.x = this.getX(i);
			_vector.y = this.getY(i);
			_vector.z = this.getZ(i);

			_vector.transformDirection(m);

			this.setXYZ(i, _vector.x, _vector.y, _vector.z);

		}

		return this;

	}

	/**
	 * @param {!Array<number>} value 
	 * @param {number=} offset 
	 */
	set(value, offset = 0) {

		this.array.set(value, offset);

		return this;

	}

	/**
	 * @param {number} index 
	 * @return {number}
	 */
	getX(index) {

		return this.array[index * this.itemSize];

	}

	/**
	 * @param {number} index 
	 * @param {number} x 
	 * @return {BufferAttribute}
	 */
	setX(index, x) {

		this.array[index * this.itemSize] = x;

		return this;

	}

	/**
	 * @param {number} index 
	 * @return {number}
	 */
	getY(index) {

		return this.array[index * this.itemSize + 1];

	}

	/**
	 * @param {number} index 
	 * @param {number} y 
	 * @return {BufferAttribute}
	 */
	setY(index, y) {

		this.array[index * this.itemSize + 1] = y;

		return this;

	}

	/**
	 * @param {number} index 
	 * @return {number}
	 */
	getZ(index) {

		return this.array[index * this.itemSize + 2];

	}

	/**
	 * @param {number} index 
	 * @param {number} z 
	 * @return {BufferAttribute}
	 */
	setZ(index, z) {

		this.array[index * this.itemSize + 2] = z;

		return this;

	}

	/**
	 * @param {number} index 
	 * @return {number}
	 */
	getW(index) {

		return this.array[index * this.itemSize + 3];

	}

	/**
	 * @param {number} index 
	 * @param {number} w
	 * @return {BufferAttribute}
	 */
	setW(index, w) {

		this.array[index * this.itemSize + 3] = w;

		return this;

	}

	/**
	 * @param {number} index 
	 * @param {number} x
	 * @param {number} y
	 * @return {BufferAttribute}
	 */
	setXY(index, x, y) {

		index *= this.itemSize;

		this.array[index + 0] = x;
		this.array[index + 1] = y;

		return this;

	}

	/**
	 * @param {number} index 
	 * @param {number} x
	 * @param {number} y
	 * @param {number} z
	 * @return {BufferAttribute}
	 */
	setXYZ(index, x, y, z) {

		index *= this.itemSize;

		this.array[index + 0] = x;
		this.array[index + 1] = y;
		this.array[index + 2] = z;

		return this;

	}

	/**
	 * @param {number} index 
	 * @param {number} x
	 * @param {number} y
	 * @param {number} z
	 * @param {number} w
	 * @return {BufferAttribute}
	 */
	setXYZW(index, x, y, z, w) {

		index *= this.itemSize;

		this.array[index + 0] = x;
		this.array[index + 1] = y;
		this.array[index + 2] = z;
		this.array[index + 3] = w;

		return this;

	}

	/**
	 * @param {function()} callback 
 	 * @return {BufferAttribute}
	 */
	onUpload(callback) {

		this.onUploadCallback = callback;

		return this;

	}


	/**
 	 * @return {BufferAttribute}
	 */
	clone() {

		return new this.constructor(this.array, this.itemSize).copy(this);

	}

	toJSON() {

		return {
			itemSize: this.itemSize,
			type: this.array.constructor.name,
			array: Array.prototype.slice.call(this.array),
			normalized: this.normalized
		};

	}

}

class Int8BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Int8Array} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Int8Array(array), itemSize, normalized);

	}

}

class Uint8BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Uint8Array} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Uint8Array(array), itemSize, normalized);

	}

}

class Uint8ClampedBufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Uint8ClampedArray} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Uint8ClampedArray(array), itemSize, normalized);

	}

}

class Int16BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Int16Array} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Int16Array(array), itemSize, normalized);

	}

}

class Uint16BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Uint16Array} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Uint16Array(array), itemSize, normalized);

	}

}

class Int32BufferAttribute extends BufferAttribute {


	/**
	 * @param {Array<number>|Float32Array|Int32Array} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Int32Array(array), itemSize, normalized);

	}

}

class Uint32BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Uint32Array} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Uint32Array(array), itemSize, normalized);

	}

}

class Float16BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Uint16Array|Float32Array|number} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Uint16Array(array), itemSize, normalized);

		Object.defineProperty(this, 'isFloat16BufferAttribute', { value: true });

	}

}

class Float32BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|number} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Float32Array(array), itemSize, normalized);

	}

}

class Float64BufferAttribute extends BufferAttribute {

	/**
	 * @param {Array<number>|Float32Array|Float64Array|number} array
	 * @param {number} itemSize
	 * @param {boolean=} normalized
	 */
	constructor(array, itemSize, normalized) {

		super(new Float64Array(array), itemSize, normalized);

	}

}

export {
	Float64BufferAttribute,
	Float32BufferAttribute,
	Float16BufferAttribute,
	Uint32BufferAttribute,
	Int32BufferAttribute,
	Uint16BufferAttribute,
	Int16BufferAttribute,
	Uint8ClampedBufferAttribute,
	Uint8BufferAttribute,
	Int8BufferAttribute,
	BufferAttribute
};
